corehq/apps/domain/tests/test_auth.py

```diff
@@ -64,3 +66,183 @@ def test_request_for_valid_domain_membership_for_non_superuser(self, *args):
 
         with patch('corehq.apps.users.models.CouchUser.is_member_of', return_value=True):
             self.assertTrue(user_can_access_domain_specific_pages(request))
+
+
+class ApiKeyFallbackTests(TestCase):
+    def test_fails_if_does_not_allow_api_keys_as_passwords(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234')
+
+        request = self._create_request(can_use_api_key=False)
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
+
+    def test_returns_user_on_successful_auth(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234')
+
+        request = self._create_request()
+
+        self.assertEqual(self.backend.authenticate(request, 'test@dimagi.com', '1234'), user)
+
+    def test_does_not_authenticate_against_inactive_keys(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', is_active=False)
+
+        request = self._create_request()
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
+
+    def test_allows_access_via_domain_restricted_keys(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', domain='test-domain')
+
+        request = self._create_request(for_domain='test-domain')
+
+        self.assertEqual(self.backend.authenticate(request, 'test@dimagi.com', '1234'), user)
+
+    def test_does_not_allow_cross_domain_access(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', domain='test-domain1')
+
+        request = self._create_request(for_domain='test-domain2')
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
+
+    def test_allows_access_to_ips_on_whitelist(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', allowed_ips=['127.0.0.1'])
+
+        request = self._create_request(for_domain='test-domain2', ip='127.0.0.1')
+
+        self.assertEqual(self.backend.authenticate(request, 'test@dimagi.com', '1234'), user)
+
+    def test_does_not_allow_ips_not_on_whitelist(self):
+        user = self._create_user('test@dimagi.com')
+        self._create_api_key_for_user(user, key='1234', allowed_ips=['127.0.0.1'])
+
+        request = self._create_request(for_domain='test-domain2', ip='5.5.5.5')
+
+        self.assertIsNone(self.backend.authenticate(request, 'test@dimagi.com', '1234'))
```

Comment:
It doesn't look like there is a test that sets `ip` on the request when the user has an empty `allowed_ips` whitelist. I think the name of that test might be `test_allows_access_to_ip_with_empty_ip_whitelist`. It seems important to test the SQL conditions (`Q(api_keys__ip_allowlist=[]) | Q(api_keys__ip_allowlist__contains=[ip])`) in that scenario.

Possibly relevant: I don't know what `get_ip(request)` returns for a request created with `self._create_request()`. It's possible it uses some reasonable default IP, in which case maybe this situation is already implicitly tested in other cases?

Reply:
Yeah, that is probably important to cover. I've kind of inverted the tests to now always test with an IP, because I assume that is the default type of request we'll be receiving. I've added in a new test to both suites to verify that you can still verify without an IP, provided you aren't using an API key with a whitelist. I don't think it's common we'll receive requests like this, but as we grab the IP from headers, I assume you can construct requests with command line tools that essentially hide your IP from us.